### PR TITLE
Remove promises from the parsing pipeline

### LINKED ISSUE #925: Remove futures/promises during tile parsing
We're currently [using futures/promises](https://github.com/mapbox/mapbox-gl-native/blob/e1ff75a47aa3893219e0a9842829e5566f26449a/src/mbgl/text/glyph_store.cpp#L261) to block tile parsing until the required glyph set has been loaded and added to the glyph atlas. However, this blocks our parsing threads. Instead, we should abort parsing the current layer and restart parsing once we have obtained the required resources.

Once this is done, we can remove the custom worker pool and instead use the worker pool provided by libuv.

### PR DESCRIPTION
**Pull request by tmpsantos**
This branch removes promises (and thus blocking) from the parsing pipeline by introducing two new concepts:

* ResourceLoader: which is an object that will manage the tile loading and emit signals to the Map thread when the data changes. A bunch of stuff that was previously on the Map thread is now living inside this object scope, most notably the worker pool. This will make our lives easier to write unit tests for parsing since we don't need anymore to bootstrap the whole stack.

* Partial tile parsing: tiles can now skip buckets that have resources pending. When the resources arrive, we just trigger another update() and the full tile gets rendered.

Fixes #925

### CODE REVIEW

**Comment by tmpsantos** [comment]
This branch is a WIP. @kkaefer @jfirebaugh please take a look.

**Comment by mb12** [comment]
@tmpsantos Does "partial tile parsing" mean that  area and line buckets will appear before symbols? Currently all buckets for a tile have to be loaded before the tile is marked as ready for rendering.

**Comment by tmpsantos** [comment]
@mb12 that is exactly the idea

**Comment by jfirebaugh** [comment]
:+1: on replacing callbacks with an observer pattern. This will bring native closer to gl-js. In gl-js, there is a clear object hierarchy: the style owns the sprite and sources, sources own tiles. Load notifications get pushed up the hierarchy: a source listens for tile load events and reemits an event of its own. The style listens for sprite and source events. A hierarchical approach has proven easy to reason about and extend -- I'd like to follow it in native.

I'm not sure about `SourceManager`. The responsibilities of this class are not clear to me, and it is not a concept that gl-js has. In gl-js, the style manages the sources. I think I'd rather do the same in native, with possibly `MapContext` also handling some of the responsibilities if this gets rebased on top of #1065.

**Comment by incanus** [comment]
Related: https://github.com/mapbox/mapbox-gl-native/issues/163

**Comment by friedbunny** [comment]
Crashed on-device with a custom `name` (i.e., not `name_en`) Streets style, looking at Japan.

```c
* thread #18: tid = 0x1fd88b, 0x002a00a6 Mapbox GL`std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > > > > >::operator[](std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [inlined] std::__1::unique_ptr<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > > >, void*>, std::__1::__hash_map_node_destructor<std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > > >, void*> > > >::reset(this=0x3569b1e5, this=0x19d3e028, __p=0x367b68c8, __p=0x19d3e028) + 28 at memory:2628, name = 'Map', stop reason = EXC_BAD_ACCESS (code=2, address=0x3569b1e5)
frame #0: 0x002a00a6 Mapbox GL`std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > > > > >::operator[](std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [inlined] std::__1::unique_ptr<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > > >, void*>, std::__1::__hash_map_node_destructor<std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > > >, void*> > > >::reset(this=0x3569b1e5, this=0x19d3e028, __p=0x367b68c8, __p=0x19d3e028) + 28 at memory:2628
frame #1: 0x002a008a Mapbox GL`std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > > > > >::operator[](std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 6 at memory:2598
frame #2: 0x002a0084 Mapbox GL`std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > > > > >::operator[](std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) [inlined] std::__1::unique_ptr<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > > >, void*>, std::__1::__hash_map_node_destructor<std::__1::allocator<std::__1::__hash_node<std::__1::__hash_value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > > >, void*> > > >::~unique_ptr(this=0x3569b1e5, this=0x0468af8c) + 4 at memory:2598
frame #3: 0x002a0080 Mapbox GL`std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, std::__1::map<std::__1::pair<unsigned short, unsigned short>, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> >, std::__1::less<std::__1::pair<unsigned short, unsigned short> >, std::__1::allocator<std::__1::pair<std::__1::pair<unsigned short, unsigned short> const, std::__1::unique_ptr<mbgl::GlyphPBF, std::__1::default_delete<mbgl::GlyphPBF> > > > > > > >::operator[](this=0x19d3e03c, __k=0x19d3e034) + 532 at unordered_map:1369
frame #4: 0x0015953e Mapbox GL`std::__1::__function::__func<std::__1::__bind<void (mbgl::Source::*)(), mbgl::Source*>, std::__1::allocator<std::__1::__bind<void (mbgl::Source::*)(), mbgl::Source*> >, void ()>::operator()() [inlined] decltype(__f=0x19d3e034, __a0=0x19d3e03c)).*fp(std::__1::forward<>(fp1))) std::__1::__invoke<void (mbgl::Source::*&)(), mbgl::Source*&, void>(void (mbgl::Source::*&&&)(), mbgl::Source*&&&) + 68 at __functional_base:380
frame #5: 0x001594fa Mapbox GL`std::__1::__function::__func<std::__1::__bind<void (mbgl::Source::*)(), mbgl::Source*>, std::__1::allocator<std::__1::__bind<void (mbgl::Source::*)(), mbgl::Source*> >, void ()>::operator()() [inlined] std::__1::__bind_return<void (mbgl::Source::*)(), std::__1::tuple<mbgl::Source*>, std::__1::tuple<>, _is_valid_bind_return<void (mbgl::Source::*)(), std::__1::tuple<mbgl::Source*>, std::__1::tuple<> >::value>::type std::__1::__apply_functor<void (__f=0x19d3e034, __bound_args=0x19d3e03c, (null)=__tuple_indices<0> at 0x0468af10, __args=0x0468af30)(), std::__1::tuple<mbgl::Source*>, 0ul, std::__1::tuple<> >(void (mbgl::Source::*&)(), std::__1::tuple<mbgl::Source*>&, std::__1::__tuple_indices<0ul>, std::__1::tuple<>&&) + 20 at functional:2023
frame #6: 0x001594e6 Mapbox GL`std::__1::__function::__func<std::__1::__bind<void (mbgl::Source::*)(), mbgl::Source*>, std::__1::allocator<std::__1::__bind<void (mbgl::Source::*)(), mbgl::Source*> >, void ()>::operator()() [inlined] std::__1::__bind_return<void (mbgl::Source::*)(), std::__1::tuple<mbgl::Source*>, std::__1::tuple<>, _is_valid_bind_return<void (mbgl::Source::*)(), std::__1::tuple<mbgl::Source*>, std::__1::tuple<> >::value>::type std::__1::__bind<void (this=0x19d3e034)(), mbgl::Source*>::operator()<>() + 14 at functional:2086
frame #7: 0x001594d8 Mapbox GL`std::__1::__function::__func<std::__1::__bind<void (mbgl::Source::*)(), mbgl::Source*>, std::__1::allocator<std::__1::__bind<void (mbgl::Source::*)(), mbgl::Source*> >, void ()>::operator()() [inlined] decltype(__f=0x19d3e034)(), mbgl::Source*>&>(fp)(std::__1::forward<>(fp0))) std::__1::__invoke<std::__1::__bind<void (mbgl::Source::*)(), mbgl::Source*>&>(std::__1::__bind<void (mbgl::Source::*)(), mbgl::Source*>&&&) + 4 at __functional_base:413
frame #8: 0x001594d4 Mapbox GL`std::__1::__function::__func<std::__1::__bind<void (mbgl::Source::*)(), mbgl::Source*>, std::__1::allocator<std::__1::__bind<void (mbgl::Source::*)(), mbgl::Source*> >, void ()>::operator(this=0x19d3e030)() + 32 at functional:1370
frame #9: 0x001389cc Mapbox GL`std::__1::function<void ()>::operator(this=0x19d3e030)() const + 116 at functional:1756
* frame #10: 0x001696b0 Mapbox GL`mbgl::TileData::reparse(this=0x19d3e028)>)::$_2::operator()() const + 44 at tile_data.cpp:84
frame #11: 0x001695b4 Mapbox GL`std::__1::__function::__func<mbgl::TileData::reparse(mbgl::Worker&, std::__1::function<void ()>)::$_2, std::__1::allocator<mbgl::TileData::reparse(mbgl::Worker&, std::__1::function<void ()>)::$_2>, void ()>::operator()() [inlined] decltype(__f=0x19d3e028)>)::$_2&>(fp)(std::__1::forward<>(fp0))) std::__1::__invoke<mbgl::TileData::reparse(mbgl::Worker&, std::__1::function<void ()>)::$_2&>(mbgl::TileData::reparse(mbgl::Worker&, std::__1::function<void ()>)::$_2&&&) + 10 at __functional_base:413
frame #12: 0x001695aa Mapbox GL`std::__1::__function::__func<mbgl::TileData::reparse(mbgl::Worker&, std::__1::function<void ()>)::$_2, std::__1::allocator<mbgl::TileData::reparse(mbgl::Worker&, std::__1::function<void ()>)::$_2>, void ()>::operator(this=0x19d3e020)() + 38 at functional:1370
frame #13: 0x001389cc Mapbox GL`std::__1::function<void ()>::operator(this=0x0468b000)() const + 116 at functional:1756
frame #14: 0x002d08a0 Mapbox GL`mbgl::Worker::afterWork(this=0x146d0070, after=mbgl::Worker::Fn at 0x0468b000)>) + 48 at worker.cpp:65
frame #15: 0x002d130e Mapbox GL`mbgl::Worker::Worker(this=0x146d016c)::$_0::operator()(std::__1::function<void ()>) const + 126 at worker.cpp:8
```

```c
frame #10: 0x001696b0 Mapbox GL`mbgl::TileData::reparse(this=0x19d3e028)>)::$_2::operator()() const + 44 at tile_data.cpp:84
   81  	            tile->endParsing();
   82  	             // `tile` is bound in this lambda to ensure that if it's the last owning pointer,
   83  	             // destruction happens on the map thread, not the worker thread.
-> 84  	            callback();
   85  	        });
   86  	}
```

**Comment by tmpsantos** [comment]
https://github.com/mapbox/mapbox-gl-native/pull/1065 landed, I will have to rewrite most of this patch.

File: src/mbgl/map/source.cpp [hunk 1]
```diff
@@ -188,7 +190,7 @@ void Source::render(Painter &painter, const StyleLayer &layer_desc) {
     gl::group group(std::string { "layer: " } + layer_desc.id);
     for (const auto& pair : tiles) {
         Tile &tile = *pair.second;
-        if (tile.data && tile.data->state == TileData::State::parsed) {
+        if (tile.data && tile.data->ready()) {
```
Comment (kkaefer) [inline src/mbgl/map/source.cpp]: Can we change this into something that indicates that state will be changed, like `setReady()`?

File: src/mbgl/map/source.cpp [hunk 2]
```diff
@@ -431,7 +460,10 @@ void Source::update(MapData& data,
         bool obsolete = std::find(retain.begin(), retain.end(), tile.id) == retain.end();
         if (!obsolete) {
             retain_data.insert(tile.data->id);
-        } else if (type != SourceType::Raster && tile.data->ready()) {
+        } else if (type != SourceType::Raster && tile.data->state == TileData::State::parsed) {
```
Comment (kkaefer) [inline src/mbgl/map/source.cpp]: Along the lines of `TileData::isReadyState()`, can we move this into a function too?

File: src/mbgl/map/sprite.hpp [hunk 3]
```diff
@@ -60,13 +67,10 @@ class Sprite : private util::noncopyable {
...
     const SpritePosition empty;
 
-    std::promise<void> promise;
-    std::future<void> future;
-    std::function<void ()> callback;
-
     Environment& env;
     Request* jsonRequest = nullptr;
     Request* spriteRequest = nullptr;
+    Observer* observer;
```
Comment (kkaefer) [inline src/mbgl/map/sprite.hpp]: Uninitialized variable

File: src/mbgl/map/resource_loader.hpp [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    // Sprite::Observer implementation.
+    virtual void onSpriteLoaded() override;
+
+private:
+    void emitTileDataChanged();
+
+    std::string accessToken_;
+    util::ptr<Sprite> sprite_;
+    Style* style_;
```
Comment (kkaefer) [inline src/mbgl/map/resource_loader.hpp]: Uninitialized pointer

File: src/mbgl/map/resource_loader.hpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    void update(MapData&, const TransformState&, GlyphAtlas&, GlyphStore&,
+                SpriteAtlas&, TexturePool&);
+
+    // FIXME: There is probably a better place for this.
+    inline util::ptr<Sprite> getSprite() const {
+        return sprite_;
+    }
+
+    // GlyphStore::Observer implementation.
+    virtual void onGlyphRangeLoaded() override;
```
Comment (kkaefer) [inline src/mbgl/map/resource_loader.hpp]: Why are they defined as `virtual`? It should suffice to define them as `override`.

File: src/mbgl/map/source.hpp [hunk 6]
```diff
@@ -116,6 +128,10 @@ class Source : public std::enable_shared_from_this<Source>, private util::noncop
     std::map<TileID, std::unique_ptr<Tile>> tiles;
     std::map<TileID, std::weak_ptr<TileData>> tile_data;
     TileCache cache;
+
+    Request* req = nullptr;
+
+    Observer* observer_;
```
Comment (kkaefer) [inline src/mbgl/map/source.hpp]: uninitialized

File: src/mbgl/map/tile_data.cpp [hunk 7]
```diff
@@ -57,7 +58,19 @@ void TileData::cancel() {
     }
 }
 
+bool TileData::mayStartParsing() {
+    return !parsing.test_and_set(std::memory_order_acquire);
```
Comment (kkaefer) [inline src/mbgl/map/tile_data.cpp]: Why is this necessary?

File: src/mbgl/map/tile_parser.cpp [hunk 8]
```diff
@@ -117,7 +119,14 @@ std::unique_ptr<Bucket> TileParser::createBucket(const StyleBucket &bucketDesc)
         } else if (bucketDesc.type == StyleLayerType::Line) {
             return createLineBucket(*layer, bucketDesc);
         } else if (bucketDesc.type == StyleLayerType::Symbol) {
-            return createSymbolBucket(*layer, bucketDesc);
+            bool needsResources;
```
Comment (kkaefer) [inline src/mbgl/map/tile_parser.cpp]: Can we replace this with a `std::tuple` return value and a `std::tie` to expand the parameters?

Comment (kkaefer) [inline src/mbgl/map/tile_parser.cpp]: Ideally, all `create*Bucket` functions will have the same return type.

File: src/mbgl/map/vector_tile_data.cpp [hunk 9]
```diff
@@ -44,34 +44,62 @@ void VectorTileData::parse() {
...
+
+Bucket* VectorTileData::getBucket(const std::string& key) const {
+    std::lock_guard<std::mutex> lock(bucketsMutex);
+
+    auto it = buckets.find(key);
+    if (it == buckets.end())
+        return nullptr;
+
+    assert(it->second);
+    return it->second.get();
```
Comment (kkaefer) [inline src/mbgl/map/vector_tile_data.cpp]: This may fail when `setBucket` overwrites an existing bucket. In that case, the existing bucket will be deallocated and a new one will be created, leaving the returned pointer invalid. `setBucket` needs to ensure that it will never overwrite an existing bucket.

File: src/mbgl/renderer/symbol_bucket.cpp [hunk 10]
```diff
@@ -120,8 +119,10 @@ std::vector<SymbolFeature> SymbolBucket::processFeatures(const GeometryTileLayer
         util::mergeLines(features);
     }
 
-    glyphStore.waitForGlyphRanges(layout.text.font, ranges);
-    sprite.waitUntilLoaded();
+    if (glyphStore.requestGlyphRangesIfNeeded(layout.text.font, ranges)) {
+        needsGlyphs_ = true;
+        return std::vector<SymbolFeature>();
```
Comment (kkaefer) [inline src/mbgl/renderer/symbol_bucket.cpp]: We can replace this with `{}`;

File: src/mbgl/map/resource_loader.hpp (hunk 5 above)
Comment (tmpsantos) [inline src/mbgl/map/resource_loader.hpp]: > Why are they defined as virtual? It should suffice to define them as override.

Just override should do it. I'm too used to pre C++11 code and prepending the virtual keyword was an inferior form annotation.

File: src/mbgl/map/tile_data.cpp (hunk 7 above)
Comment (tmpsantos) [inline src/mbgl/map/tile_data.cpp]: > Why is this necessary?

Because tiles in the `partial` state are always parsed, this will prevent the same tile from being parsed in parallel by one or more workers.

**Comment by kkaefer** [comment]
\o/
